src/sampling.py

```diff
@@ -22,6 +22,51 @@ def mnist_iid(dataset, num_users):
         all_idxs = list(set(all_idxs) - dict_users[i])
     return dict_users
 
+def split_dirichlet(dataset, num_users, is_cfar, beta=0.5):
+    """
+    Sample non-I.I.D client data from an arbitary dataset.
+    Samples it based on this paper: 10.48550/ARXIV.1905.12022
+    :param dataset: The dataset
+    :param num_users: The number of clients
+    :param is_cfar: Whether the dataset is cfar (bool). This is a stupid solution but is necessary since for some reason,
+    their parameter name is different.
+    :return: dict mapping client id to idxs for training
+    """
+    idxs = np.arange(len(dataset))
+    labels = np.array(dataset.targets) if is_cfar else dataset.train_labels.numpy()
+    uniq_labels = np.unique(labels)
+    dict_users = {i: np.array([]) for i in range(num_users)}
+    idxs_labels = np.vstack((idxs, labels))
+    idxs_labels = idxs_labels[:, idxs_labels[1, :].argsort()]
+    idxs_labels = idxs_labels.transpose()
+
+    assert len(idxs_labels[1]) == 2 # sanity check
+
+    for label in uniq_labels:
+        relevant_idxs = idxs_labels[(idxs_labels[:, 1] == label)][:,0].T
+        proportions = np.random.dirichlet(np.full(num_users, beta))
+        splits = split_by_ratio(relevant_idxs, proportions)
+        for idx, split in enumerate(splits):
+            dict_users[idx] = np.concatenate([dict_users[idx], split])
+
+    return dict_users
```

Comment:
The logic looks correct. Have you tested it manually?

Reply:
Yes. Here's a split on 10 clients using MNIST. 

0: {0: 413, 1: 383, 2: 422, 3: 692, 4: 1229, 5: 2336, 6: 686, 7: 1073, 8: 169, 9: 21}, 
1: {0: 2757, 1: 683, 2: 331, 3: 377, 4: 1077, 5: 599, 6: 27, 7: 727, 8: 120}, 
2: {0: 496, 1: 4, 2: 1145, 3: 180, 4: 298, 5: 445, 6: 44, 7: 3, 8: 881, 9: 10}, 
3: {0: 635, 1: 2, 2: 90, 3: 164, 4: 521, 5: 74, 6: 599, 7: 84, 8: 213, 9: 1850}, 
4: {0: 28, 1: 841, 2: 269, 3: 454, 4: 4, 5: 563, 6: 25, 7: 667, 8: 1063, 9: 2376}, 
5: {0: 215, 2: 605, 3: 2402, 4: 1795, 5: 412, 6: 162, 7: 25, 8: 1076, 9: 461}, 
6: {0: 1, 1: 1819, 2: 227, 3: 1497, 4: 97, 5: 116, 6: 1166, 7: 614, 8: 1430, 9: 699}, 
7: {0: 51, 1: 579, 2: 2704, 3: 128, 4: 21, 5: 109, 6: 44, 7: 191, 8: 774, 9: 344}, 
8: {0: 126, 1: 2240, 2: 3, 3: 185, 4: 771, 5: 56, 6: 2030, 7: 62, 8: 63, 9: 65}, 
9: {0: 1201, 1: 191, 2: 162, 3: 52, 4: 28, 5: 711, 6: 1135, 7: 2819, 8: 61, 9: 123}}